
<launch>
   
   <!-- Example to run rgbd datasets:
   $ wget http://vision.in.tum.de/rgbd/dataset/freiburg3/rgbd_dataset_freiburg3_long_office_household.bag
   $ rosbag decompress rgbd_dataset_freiburg3_long_office_household.bag

   $ roslaunch rtabmap_ros rgbdslam_datasets.launch feature=6
   $ rosbag play -.-clock rgbd_dataset_freiburg3_long_office_household.bag
   -->
   
   <param name="use_sim_time" type="bool" value="True"/>
     
   <!-- Choose visualization -->
   <arg name="rviz" default="true" />
   <arg name="rtabmapviz" default="false" /> 
 
  <!-- TF FRAMES -->
  <node pkg="tf" type="static_transform_publisher" name="world_to_map" 
    args="0.0 0.0 0.0 0.0 0.0 0.0 /world /map 100" />
    
      
  <group ns="rtabmap">
  
    <!-- Odometry -->
    <node pkg="rtabmap_ros" type="rgbd_odometry" name="rgbd_odometry" output="screen">
      <remap from="rgb/image"       to="/camera/rgb/image_color"/>
      <remap from="depth/image"     to="/camera/depth/image"/>
      <remap from="rgb/camera_info" to="/camera/rgb/camera_info"/>
      <remap from="odom" to="vis_odom"/>
	  
      <param name="Odom/Strategy" type="string" value="0"/>      <!-- 0=Frame-to-Map, 1=Frame-to-KeyFrame -->
      <param name="Vis/CorType" type="string" value="0"/>        <!-- 0=features matching 1=Optical Flow -->
      <param name="Vis/EstimationType" type="string" value="0"/> <!-- 0=3D->3D, 1=3D->2D (PnP) -->
      <param name="Odom/FillInfoData" type="string" value="$(arg rtabmapviz)"/>
      <param name="Vis/MaxDepth" type="string" value="4"/>
      <param name="Vis/CorNNDR" type="string" value="0.6"/>
      <param name="Odom/ResetCountdown" type="string" value="15"/> 
      <param name="Odom/KeyFrameThr" type="string" value="0.5"/>

      <param name="odom_frame_id" type="string" value="vis_odom"/>
      <param name="frame_id" type="string" value="kinect"/>
      <param name="publish_tf" type="bool" value="false"/>
      <param name="queue_size" type="int" value="30"/>
      <param name="wait_for_transform" type="bool" value="true"/>
      <param name="ground_truth_frame_id" type="string" value="world"/>
    </node>
    <!-- rename the child frame of odometry -->
    <node name="odom_msg_to_tf" pkg="rtabmap_ros" type="odom_msg_to_tf">
      <param name="frame_id" type="string" value="kinect_est"/>
    </node>
    
    <!-- Visual SLAM -->
    <!-- args: "delete_db_on_start" and "udebug" -->
    <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen" args="--delete_db_on_start">
      <param name="subscribe_depth" type="bool" value="true"/>
      <param name="subscribe_laserScan" type="bool" value="false"/>
      
      <param name="Rtabmap/StartNewMapOnLoopClosure" type="string" value="true"/>
      <param name="Vis/EstimationType" type="string" value="0"/>
      <param name="Vis/MaxDepth" type="string" value="4"/>
      <param name="RGBD/LoopClosureReextractFeatures" type="string" value="false"/>
      <param name="Mem/RawDescriptorsKept" type="string" value="true"/>
      <param name="Kp/DetectorStrategy" type="string" value="0"/>

      <param name="frame_id" type="string" value="kinect"/>
      <param name="ground_truth_frame_id" type="string" value="world"/> 
	
      <remap from="rgb/image" to="/camera/rgb/image_color"/>
      <remap from="depth/image" to="/camera/depth/image"/>
      <remap from="rgb/camera_info" to="/camera/rgb/camera_info"/>
      <remap from="odom" to="vis_odom"/>

      <param name="queue_size" type="int" value="30"/>
    </node>
    
    <!-- Visualisation  -->
    <node if="$(arg rtabmapviz)" pkg="rtabmap_ros" type="rtabmapviz" name="rtabmapviz" args="-d $(find rtabmap_ros)/launch/config/rgbd_gui.ini" output="screen">
      <param name="subscribe_depth" type="bool" value="true"/>
      <param name="subscribe_laserScan" type="bool" value="false"/>
      <param name="subscribe_odom_info" type="bool" value="true"/>
      <param name="queue_size" type="int" value="30"/>
      
      <param name="frame_id" type="string" value="kinect"/>
    
      <remap from="rgb/image" to="/camera/rgb/image_color"/>
      <remap from="depth/image" to="/camera/depth/image"/>
      <remap from="rgb/camera_info" to="/camera/rgb/camera_info"/>
      <remap from="odom" to="vis_odom"/>
    </node>
        
  </group>
  
  <node if="$(arg rviz)" pkg="rviz" type="rviz" name="rviz" args="-d $(find rtabmap_ros)/launch/config/rgbdslam_datasets.rviz"/>
  <node if="$(arg rviz)" pkg="nodelet" type="nodelet" name="points_xyzrgb" args="standalone rtabmap_ros/point_cloud_xyzrgb">
    <remap from="rgb/image"       to="/camera/rgb/image_color"/>
    <remap from="depth/image"     to="/camera/depth/image"/>
    <remap from="rgb/camera_info" to="/camera/rgb/camera_info"/>
    <remap from="cloud"           to="voxel_cloud" />
    
    <param name="queue_size" type="int" value="10"/>
    <param name="decimation" type="double" value="4"/>
  </node>

</launch>
